ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c

```diff
@@ -0,0 +1,157 @@
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/alarm/touch/TouchAlarm.h"
+#include "shared-bindings/microcontroller/__init__.h"
+
+#include "esp_sleep.h"
+#include "peripherals/touch.h"
+#include "supervisor/esp_port.h"
+
+static volatile bool woke_up = false;
+static touch_pad_t touch_channel = TOUCH_PAD_MAX;
+
+void common_hal_alarm_touch_touchalarm_construct(alarm_touch_touchalarm_obj_t *self, const mcu_pin_obj_t *pin) {
+    if (pin->touch_channel == TOUCH_PAD_MAX) {
+        mp_raise_ValueError(translate("Invalid pin"));
+    }
+    claim_pin(pin);
+    self->pin = pin;
+}
+
+mp_obj_t alarm_touch_touchalarm_get_wakeup_alarm(const size_t n_alarms, const mp_obj_t *alarms) {
+    // First, check to see if we match any given alarms.
+    for (size_t i = 0; i < n_alarms; i++) {
+        if (MP_OBJ_IS_TYPE(alarms[i], &alarm_touch_touchalarm_type)) {
+            return alarms[i];
+        }
+    }
+
+    alarm_touch_touchalarm_obj_t *alarm = m_new_obj(alarm_touch_touchalarm_obj_t);
+    alarm->base.type = &alarm_touch_touchalarm_type;
+    alarm->pin = NULL;
+
+    // Map the pin number back to a pin object.
+    for (size_t i = 0; i < mcu_pin_globals.map.used; i++) {
+        const mcu_pin_obj_t* pin_obj = MP_OBJ_TO_PTR(mcu_pin_globals.map.table[i].value);
+        if (pin_obj->touch_channel == touch_channel) {
+            alarm->pin = mcu_pin_globals.map.table[i].value;
+            break;
+        }
+    }
+
+    return alarm;
+}
+
+// This is used to wake the main CircuitPython task.
+void touch_interrupt(void *arg) {
+    (void) arg;
+    woke_up = true;
+    BaseType_t task_wakeup;
+    vTaskNotifyGiveFromISR(circuitpython_task, &task_wakeup);
+    if (task_wakeup) {
+        portYIELD_FROM_ISR();
+    }
+}
+
+void alarm_touch_touchalarm_set_alarm(const bool deep_sleep, const size_t n_alarms, const mp_obj_t *alarms) {
+    bool touch_alarm_set = false;
+    alarm_touch_touchalarm_obj_t *touch_alarm = MP_OBJ_NULL;
+
+    for (size_t i = 0; i < n_alarms; i++) {
+        if (MP_OBJ_IS_TYPE(alarms[i], &alarm_touch_touchalarm_type)) {
+            if (!touch_alarm_set) {
+                if (deep_sleep) {
+                    touch_alarm = MP_OBJ_TO_PTR(alarms[i]);
+                    touch_alarm_set = true;
+                } else {
+                    mp_raise_NotImplementedError(translate("TouchAlarm not available in light sleep"));
```

Comment:
Does the interrupt work during light sleep?

Reply:
Thanks for pointing this out. IDF implementation of `TouchWake` doesn't work for `LightSleep`. 
I forgot to test this with mechanics of `pretend-to-sleep`... wake on interrupt from light sleep is working now.